crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -391,16 +393,40 @@ impl<'db> Class<'db> {
                     );
                 }
                 ClassBase::Class(class) => {
-                    if let member @ SymbolAndQualifiers(Symbol::Type(_, _), _) =
+                    if let member @ SymbolAndQualifiers(Symbol::Type(ty, boundness), _) =
                         class.own_instance_member(db, name)
                     {
-                        return member;
+                        if boundness == Boundness::Bound {
+                            if union.is_empty() {
+                                // Short-circuit, no need to allocate inside the union builder
+                                return member;
+                            }
+
+                            return SymbolAndQualifiers(
+                                Symbol::bound(union.add(ty).build()),
+                                TypeQualifiers::empty(),
```

Comment:
Here and below `TypeQualifiers::empty()` doesn't feel right, but maybe it doesn't make sense to address this until we have support for more type qualifiers? We are throwing away type qualifiers from the `match` above.

Reply:
Yes, thanks. Added a test case, preliminary handling of type qualifiers (without diagnostics for conflicting qualifiers), and a TODO.